src/components/CustomSelect.jsx

```diff
@@ -24,37 +27,72 @@ function CustomSelect({ options, selected, onSelect }) {
     };
   });
 
+  useEffect(() => {
+    const liHeight = liRefs.current[focusedIndex]?.offsetHeight;  //total height of li element
+    const scrollTop = ulRef.current?.scrollTop;                   //shows the scroll position of ul element
+    const viewport = scrollTop + ulRef.current?.offsetHeight;     //sum of scroll position of ul and it's total height
+    const liOffset = liHeight * focusedIndex;                     //total height of li element depending on the index
+    
+    console.log('focusedIndex', focusedIndex, ' viewport', viewport, ' liOffset', liOffset);
+
+    if (ulRef.current && (liOffset < scrollTop || liOffset + liHeight > viewport)) {
+      ulRef.current.scrollTop = liOffset;
+    }
+  }, [focusedIndex]);
+
+  function handleKeyDown(event) {
+    if (!isOpen) return;
+    event.preventDefault();
+    switch (event.key) {
+      case 'ArrowUp':
+        setFocusedIndex((prevIndex) => (prevIndex <= 0 ? options.length - 1 : prevIndex - 1));
```

Comment:
arrows shouldn't be infinite by default 😁 they shouldn't go to the first element after the last one

to check if your components simulates the native behavior you can make the native one next to it 👌 
Here should be also `space` and `esc` keys included 😄 

<img width="442" alt="Zrzut ekranu 2023-06-15 o 11 36 34" src="https://github.com/darionara/custom-select/assets/30324689/0ec1ca89-3c7b-4d86-bab3-6f882d4a9782">

Reply:
✅